pkg/webhook/certs.go

```diff
@@ -0,0 +1,484 @@
+	}
+
+	crt, err := x509.ParseCertificate(b.Bytes)
+	if err != nil {
+		return false, errors.Wrap(err, "parsing cert")
+	}
+	_, err = crt.Verify(x509.VerifyOptions{
+		DNSName:     dnsName,
+		Roots:       pool,
+		CurrentTime: at,
+	})
+	if err != nil {
+		return false, errors.Wrap(err, "verifying cert")
+	}
+	return true, nil
+}
+
+// controller code for making sure the CA cert on the
+// validatingwebhookconfiguration doesn't get clobbered
+
+var _ handler.Mapper = &mapper{}
+
+type mapper struct {}
+
+func (m *mapper) Map(object handler.MapObject) []reconcile.Request {
+	if object.Meta.GetNamespace() != vwhKey.Namespace {
+		return nil
+	}
+	if object.Meta.GetName() != vwhKey.Name {
+		return nil
+	}
+	return []reconcile.Request{{NamespacedName: secretKey}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func addController(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(("validating-webhook-controller"), mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the provided ValidatingWebhookConfiguration
+	s := &corev1.Secret{}
+	if err := c.Watch(&source.Kind{Type: s}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(vwhGVK)
+	mapper := &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper{}}
+	if err := c.Watch(&source.Kind{Type: vwh}, mapper); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileVWH{}
+
+// ReconcileVWH reconciles a validatingwebhookconfiguration, making sure it
+// has the appropriate CA cert
+type ReconcileVWH struct {
+	client client.Client
+	scheme *runtime.Scheme
+	ctx context.Context
+}
+
+// Reconcile reads that state of the cluster for a validatingwebhookconfiguration
+// object and makes sure the most recent CA cert is included
+func (r *ReconcileVWH) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if request.NamespacedName != secretKey {
+		return reconcile.Result{}, nil
+	}
+	secret := &corev1.Secret{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, secret); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(vwhGVK)
+	if err := r.client.Get(r.ctx, vwhKey, vwh); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
```

Comment:
Should requeue be `Requeue: true`

Reply:
fixed

Reply:
This works, but note that Requeue: true is unnecessary when you return an error. The caller always requeues if an error is returned.